### PR TITLE
Rust/leap add sample description

### PR DESCRIPTION
**Pull request by ptillemans**
I added some sample description for the *leap* exercise in the rust track. It is a non core exercise but I wanted to reuse @yawpitch work for python and also experiment and discuss on slack before starting the *real* work.

- I added a .gitignore on the rust track for avoiding stuff to be appearing in the repo.
- I converted the *README.md* from pythonese to rustish
- I added sample code with benchmarks to back up the claims in the READMP

### CODE REVIEW

File: tracks/rust/exercises/leap/README.md [hunk 1]
```diff
@@ -0,0 +1,63 @@
+### Reasonable solutions
+
+```rust
+pub fn is_leap_year(year):
+    year % 4 == 0 && (year % 100 != 0 || year % 400 == 0)
+```
+
+### Common suggestions
+- there are just two cases that return True:
+  - a year is a multiple of 4 *and not** 100
```
Comment (Emerentius) [inline tracks/rust/exercises/leap/README.md]: is this supposed to be cursive or bold? It starts with 1 star, ends with 2

File: tracks/rust/exercises/leap/README.md [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+test tests::leap1_bench ... bench:       1,644 ns/iter (+/- 103)
+test tests::leap2_bench ... bench:       1,641 ns/iter (+/- 182)
+test tests::leap3_bench ... bench:       1,628 ns/iter (+/- 105)
+test tests::leap4_bench ... bench:       2,637 ns/iter (+/- 312)
+
+test result: ok. 0 passed; 0 failed; 0 ignored; 4 measured; 0 filtered out
+```
+
+This clearly shows that:
+- binding sub expressions is zero-cost
```
Comment (Emerentius) [inline tracks/rust/exercises/leap/README.md]: This is an example of the optimizer changing evaluation order. Does this really hold in general?

Comment (ptillemans) [reply tracks/rust/exercises/leap/README.md]: It does not change evaluation order. The evaluation order is defined by the precedence rules in the expression. The compiler sees the bindings a constants and can use constant propagation to move the expression attached to it to where it is used and inline the evaluation. This is quite generic.

Comment (ptillemans) [reply tracks/rust/exercises/leap/README.md]: see also https://llvm.org/docs/Passes.html#scalar-evolution-scalar-evolution-analysis

Comment (Emerentius) [reply tracks/rust/exercises/leap/README.md]: Rust's semantics are not defined by LLVM. Are you basing your statements on a binding model defined by Rust, in the reference maybe?

**Comment by ptillemans** [comment]
Of course the semantics are not defined by LLVM, it is Rust which generates
the IM code which contains the encoded semantics.

That being said, let's look at what the reference has to say about this:
https://doc.rust-lang.org/reference/variables.html

> A *local variable* (or *stack-local* allocation) holds a value directly,
allocated within the stack's
> memory. The value is a part of the stack frame.
> ...
> Local variables are not initialized when allocated. Instead, the entire
frame worth of local variables
> are allocated, on frame-entry, in an uninitialized state. Subsequent
statements within a function may
> or may not initialize the local variables. Local variables can be used
only after they have been
> initialized; this is enforced by the compiler.

note also what it does not say: it is nowhere stated that the variables are
initialized when they are declared, just that they are initialized before
they will be used.

This gives LLVM leeway to apply it's optimization logic, as long as the
value is calculated before it is being used.

In debug versions they will be calculated in the stack frame, but I would
bet that in the optimized release versions these are optimized out.

Let's look at the assembly for both versions:

```asm
__ZN4leap13is_leap_year117h70506ca920b2d885E:
        .cfi_startproc
        pushq   %rbp
        .cfi_def_cfa_offset 16
        .cfi_offset %rbp, -16
        movq    %rsp, %rbp
        .cfi_def_cfa_register %rbp
        testb   $3, %dil
        jne     LBB0_1
        ...
LBB0_1:
        xorl    %eax, %eax
        popq    %rbp
        retq
...
__ZN4leap13is_leap_year217h23ef1c74ccbfde3aE:
        .cfi_startproc
        pushq   %rbp
        .cfi_def_cfa_offset 16
        .cfi_offset %rbp, -16
        movq    %rsp, %rbp
        .cfi_def_cfa_register %rbp
        testb   $3, %dil
        je      LBB1_3
        xorl    %eax, %eax
        popq    %rbp
        retq
LBB1_3:
```

you see for both variants, *exactly* the same code is generated (except for
the 'jne' and 'je').

**Comment by Emerentius** [comment]
If the expressions that assign to those variables had side-effects or were opaque, they would not be reordered. Neither would they be reordered in all the same ways if the computation is expensive and the result used multiple times.
The order of the statements (or expressions) here defines the order of operations. Binding the results to variables and conditionally using them necessarily includes changing the order of operations. The optimizer can reorder them as it wishes, *if* it can show them to be pure, as it helpfully does in this case.

Making a general statement like
>- binding sub expressions is zero-cost

seems like too strong of an inference to make from this small example

**Comment by ptillemans** [comment]
Aahhh, now I see your point. Indeed when there are side effects the compiler can make very little assumptions and will have to do whatever guarantees necessary to maintain the semantics. I'll update that.

File: tracks/rust/exercises/leap/README.md (hunk 1 above)
Comment (yawpitch) [reply tracks/rust/exercises/leap/README.md]: When I wrote it in the Python version it was meant to be italics.

**Comment by ErikSchierboom** [comment]
@ptillemans @Emerentius What is the status of this PR? Is there still work to be done, or is it ready to be merged?

**Comment by ptillemans** [comment]
I have no pending work on this. I would rather see it merged so it can be improved upon.

**Comment by ErikSchierboom** [comment]
Merged!

**Pull request merged**